dht/token.cc

```diff
@@ -79,35 +79,10 @@ sstring token::to_sstring() const {
     return seastar::to_sstring<sstring>(long_token(*this));
 }
 
-// Assuming that x>=y, return the positive difference x-y.
-// The return type is an unsigned type, as the difference may overflow
-// a signed type (e.g., consider very positive x and very negative y).
-template <typename T>
-static std::make_unsigned_t<T> positive_subtract(T x, T y) {
-        return std::make_unsigned_t<T>(x) - std::make_unsigned_t<T>(y);
-}
-
 token token::midpoint(const token& t1, const token& t2) {
-    auto l1 = long_token(t1);
-    auto l2 = long_token(t2);
-    int64_t mid;
-    if (l1 <= l2) {
-        // To find the midpoint, we cannot use the trivial formula (l1+l2)/2
-        // because the addition can overflow the integer. To avoid this
-        // overflow, we first notice that the above formula is equivalent to
-        // l1 + (l2-l1)/2. Now, "l2-l1" can still overflow a signed integer
-        // (e.g., think of a very positive l2 and very negative l1), but
-        // because l1 <= l2 in this branch, we note that l2-l1 is positive
-        // and fits an *unsigned* int's range. So,
-        mid = l1 + positive_subtract(l2, l1)/2;
-    } else {
-        // When l2 < l1, we need to switch l1 and and l2 in the above
-        // formula, because now l1 - l2 is positive.
-        // Additionally, we consider this case is a "wrap around", so we need
-        // to behave as if l2 + 2^64 was meant instead of l2, i.e., add 2^63
-        // to the average.
-        mid = l2 + positive_subtract(l1, l2)/2 + 0x8000'0000'0000'0000;
-    }
+    uint64_t l1 = long_token(t1);
+    uint64_t l2 = long_token(t2);
+    int64_t mid = l1 + (l2 - l1)/2;
```

Comment:
Let's try two examples:

1. l1=1, l2=-1

then mid = 1 + (0xffff'ffff'ffff'ffff - 1) / 2 = 1 + 0x7fff'ffff'ffff'ffff = 0x8000'0000'0000'0000, matches the old code.

2. l1=0x7fff'ffff'fffff'ffff, l2=-0x7ffff'ffff'ffff'ffff

then mid = 0x7fff'ffff'fffff'ffff +  (0x8000'0000'0000'0001 - 0x7fff'ffff'fffff'ffff) / 2 = 0x8000'0000'0000'0000, again matches the old code.

Reply:
btw, I think the only remaining user of midpoint() is in token_test.